gradle-plugin/src/main/java/com/contrastsecurity/gradle/plugin/ResolveAgentTask.java

```
@@ -0,0 +1,106 @@
+    this.fileString = fileString;
+  }
+
+  public void setAgent(File agent) {
+    this.agent = agent;
+  }
+
+  @TaskAction
+  void resolveAgent() {
+    System.out.println("running RESOLVE AGENT");
+    final ContrastSDK sdk = ContrastSDKService.getInstance(config);
+    agent = retrieveAgent(sdk, getFileString(), config.getOrgUuid(), getProject());
+  }
+
+  /** Use ContrastSDK to download agent and return the path where agent jar is stored */
+  @VisibleForTesting
+  public static File retrieveAgent(
+      final ContrastSDK connection,
+      final String jarPath,
+      final String uuid,
+      final Project project) {
+
+    final Logger logger = project.getLogger();
+    // Initially attempt to get agent from the previously configured location
+    if (jarPath != null) {
+      final Path agent = Paths.get(jarPath);
+      if (!Files.exists(agent)) {
+        logger.debug("Unable to find java agent at " + jarPath);
+      } else {
+        logger.debug("Agent provided via configuration retrieved");
+        return new File(agent.toString());
+      }
+    }
+
+    logger.debug("Attempting to retrieve agent from TeamServer");
+    // If no jar is provided, and no jarpath configured, attempt to retrieve the agent from TS
+    final byte[] bytes;
+    Path downloadedAgent;
+    try {
+      bytes = connection.getAgent(AgentType.JAVA, uuid);
+
+      // Save the jar to the 'target' directory
+      final Path target = Paths.get(project.getProjectDir().getPath());
+
+      try {
+        Files.createFile(target);
+      } catch (FileAlreadyExistsException e) {
+        logger.debug("Project dir already exists");
+      }
+
+      downloadedAgent = target.resolve(jarPath);
+
+      Files.write(downloadedAgent, bytes, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+
+    } catch (RuntimeException | IOException e) {
+      throw new GradleException("Failed to download java agent from the Contrast api: " + e);
```

I wonder if we should really throw an exception on this one...  I'm not sure we want to block a customer from running tests just because we weren't able to obtain an agent.

I think we should have some kind of fallback that emits a large message in the Gradle logs that the agent isn't being run or something.  I don't know how others feel about this but as written I'm pretty sure this will dump and block all testing.

- [ ] Verify that I'm right about that

---

The task needed to run this is separate from Test Tasks right now. Running `contrastCheck` runs all the tests and explicitly expects the agent to be attached. I think I've been swapping back and forth between auto connecting contrast with the plugin applied vs having a task to do it. This way makes more sense to me right now but I am open to discussing it

---

If it's being run separately then this is fine.  It's only an issue if this was an 'always run' situation. I'd make sure this doesn't make things croak however like just refreshing Gradle for example.

---

You made me go back and check and actually you were right. The plugin configures tests to have the agent args always, because this happens in the plugin itself rather than a task. I'm conflicted again on how this should work. I think I'm going to experiment briefly with configuring the `Test` tasks within the `contrastCheck` task. I think this is antithetical to what Gradle wants with Lifecycle tasks but 🤷

---

https://contrast.atlassian.net/browse/JAVA-8962